### ISSUE TITLE
pytest : No module named 'Cython' Found in setup.py 

### ISSUE THREAD

**Issue by anandxkumar** [radis/radis#262]
### 🐛 Describe the bug

After #234, `pytest` fails on the local build.
While installing dependencies using `pip install -e .[dev] ` on a new virtual environment or existing, following error occurs -
```
Obtaining file:///home/pipebomb/radis
    ERROR: Command errored out with exit status 1:
     command: /home/pipebomb/anaconda3/envs/check_a/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/pipebomb/radis/setup.py'"'"'; __file__='"'"'/home/pipebomb/radis/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-0frumtj7
         cwd: /home/pipebomb/radis/
    Complete output (5 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/pipebomb/radis/setup.py", line 39, in <module>
        from Cython.Distutils import build_ext
    ModuleNotFoundError: No module named 'Cython'
    ----------------------------------------
WARNING: Discarding file:///home/pipebomb/radis. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.

```
While running `pytest` on develop branch clone, `test_arrays.py `fails with the following traceback -
<details>
<summary>Traceback</summary>
______________________________ test_cython_add_at ______________________________

args = (), kwargs = {}

    def test_cython_add_at(*args, **kwargs):
        """
        Compare the workings of the Cython compiled add_at() function
        versus the numpy add.at() function with bogus data.
        """
    
        # First check if Python was able to import the Cython version of add at:
>       from radis_cython_extensions import add_at as cython_add_at
E       ModuleNotFoundError: No module named 'radis_cython_extensions'

radis/test/misc/test_arrays.py:165: ModuleNotFoundError
__________________________ test_cython_add_at_spectra __________________________

args = (), kwargs = {}, wstep = 0.005, wmin = 2100, wmax = 2200, T = 1200
p = 0.1, sf = <radis.lbl.factory.SpectrumFactory object at 0x7f2f20538a10>
s_numpy = <radis.spectrum.spectrum.Spectrum object at 0x7f2f663048d0>
@py_assert1 = {'broadening_max_width': 10, 'cutoff': 1e-27, 'broadening_method': 'fft', 'optimization': 'simple', 'db_use_cached': T...0788016573, 'add_at_used': 'numpy', 'include_neighbouring_lines': True, 'folding_thresh': 1e-06, 'zero_padding': 22002}
@py_assert3 = 'numpy', @py_assert6 = 'cython'

    def test_cython_add_at_spectra(*args, **kwargs):
        """
        Test if the Cython add_at() produces the same spectra as
        with numpy add.at().
        """
    
        setup_test_line_databases()  # add HITRAN-CO-TEST in ~/.radis if not there
    
        # Conditions
        wstep = 0.005
        wmin = 2100  # cm-1
        wmax = 2200  # cm-1
    
        T = 1200  # K
        p = 0.1  # bar
    
        sf = SpectrumFactory(
            wavenum_min=wmin,
            wavenum_max=wmax,
            mole_fraction=1,
            path_length=1,  # doesnt change anything
            wstep=wstep,
            pressure=p,
            isotope="1",
            verbose=False,
            warnings={
                "MissingSelfBroadeningWarning": "ignore",
                "NegativeEnergiesWarning": "ignore",
                "HighTemperatureWarning": "ignore",
                "OutOfRangeLinesWarning": "ignore",
                "GaussianBroadeningWarning": "ignore",
                "CollisionalBroadeningWarning": "ignore",
                "AccuracyWarning": "ignore",
            },
        )
        sf.load_databank("HITRAN-CO-TEST")
    
        sf.use_cython = False
        s_numpy = sf.eq_spectrum(Tgas=T, name="numpy")
        s_numpy.apply_slit(0.5, "nm")
        assert sf.params.add_at_used == "numpy"
    
        sf.use_cython = True
        s_cython = sf.eq_spectrum(Tgas=T, name="cython")
        s_cython.apply_slit(0.5, "nm")
>       assert sf.params.add_at_used == "cython"
E       AssertionError: assert 'numpy' == 'cython'
E         - cython
E         + numpy

radis/test/misc/test_arrays.py:239: AssertionError
</details>

The same runs on Travis without any issue. 

### 💡 Possible solutions

- Create a virtual environment using `environment.yml`  and then execute `pip install -e .[dev] ` to install all dependencies including Cython, then pytest passes all tests.

- `from Cython.Distutils import build_ext` is executed before the installation of Cython in `setup.py`.
Make sure Cython is installed before Cython related files are imported or executed.

- Manually install Cython using `pip install Cython` (Version specific, else Assertion Error exists, version conflict b/w numpy and cython)?

**Comment by erwanp** [comment]
This is not critical as the main code always has a no-Cython option. This specific test fails because it explicitely test the Cython implementation. 

I don't think installting Cython with pip will do a good job. 
Installing from the environment.yml looks like the best option. Short term : we could explicitely add this in the Developer Guide.

An alternative is to start deploying Radis from Conda. This would also solve the [pytables-dependency problem](https://radis-radiation.slack.com/archives/C9F7X0ZAT/p1619264311059600) reported on the Slack channel. Unfortunately, I've never deployed a Conda package !

**Comment by erwanp** [comment]
@pkj-m if you can have a look too, being one of the Radis Cython experts ;) !

**Comment by erwanp** [comment]
Hello. Forget my first answer, I only understood that test_arrays.py did not pass (which is not critical), but I did not see that the installation `pip install -e .` itself fail on a non-cython environment (which is critical, because we cannot switch to the runtime pure-python version anymore !).

This currently makes the build fail on Readthedocs.io. 
```
Processing /home/docs/checkouts/readthedocs.org/user_builds/radis/checkouts/develop
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
    ERROR: Command errored out with exit status 1:
     command: /home/docs/checkouts/readthedocs.org/user_builds/radis/envs/develop/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-ztk2ssqk/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-ztk2ssqk/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-_rah2412
         cwd: /tmp/pip-req-build-ztk2ssqk/
    Complete output (5 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-ztk2ssqk/setup.py", line 39, in <module>
        from Cython.Distutils import build_ext
    ModuleNotFoundError: No module named 'Cython'
```

This happens because starting from #234 Cython is imported directly in `setup.py` even before other packages are installed. 
```
from Cython.Distutils import build_ext 
```

@dcmvdbekerom some web-research pointed me to **sqlalchemy** as a reference of how to deploy C-extensions. In their [setup.py](https://github.com/sqlalchemy/sqlalchemy/blob/master/setup.py)  I noted that they import build_ext from distutils. Do you have an idea if it could work for us?  
```
from distutils.command.build_ext import build_ext
```

---

Easy way to test if the install works : RADIS install should be working on a minimal, bare installation. 
```
conda create --name testenv
conda activate testenv
cd radis
pip install .[dev]
```
Even if not recommended for performance, it would work before #234. And the official documentation still recommend `pip install radis`.

**Comment by dcmvdbekerom** [comment]
I tried the test in a new conda environment but for me installing radis this way without any issues.

Anyway, in https://github.com/radis/radis/pull/266 I have now moved all cython references to within a try..except block, this way it should be skipped if cython is not installed and not throw an exception.

**Comment by anandxkumar** [comment]
It is probably because Cython is already installed on your system, so whenever Cython has to be installed on a new environment it will try to copy the cache instead of throwing the error. It should fail on any build online having similar environment because then there is no cache, just as it happened in building `Readthedocs.io` mentioned by Erwan above. 
Hopefully, #266  will fix the issue.

**Comment by dcmvdbekerom** [comment]
@anandxkumar I just merged https://github.com/radis/radis/pull/266, can you check if this solved the issue for you?

**Comment by anandxkumar** [comment]
@dcmvdbekerom, the issue has been fixed after #266, because I'm now able to install all dependencies on a new non-cython environment using `pip install -e .[dev]`  and also `pytest` passed. It can also be verified by the fact that the `Readthedocs.io` has got updated.
Thanks @dcmvdbekerom for the fix :+1:
